### ISSUE TITLE
False positive on `UnnecessaryEventHandlerParameter`

### ISSUE THREAD

**Issue by BraisGabin** [appKODE/detekt-rules-compose#13]
Given this code:

```kotlin
internal sealed class State {
    object Loading : State()
    data class Data(val id: String) : State()
}

@Composable
internal fun MyButton(
    state: State,
    onClick: (String) -> Unit,
) {
    when (state) {
        State.Loading -> Text("Loading")
        is State.Data -> Button(onClick = { onClick(state.id) }) { Text("Click here") }
    }
}
```

`UnnecessaryEventHandlerParameter` raises an issue. But in this case the parameter is necessary because `state.id` is accesible because of the smart cast that you don't have in the function that call this component.

**Comment by dimsuz** [comment]
Ugh, right. Not sure how to solve this without converting this rule to use type resolution (to check that `State` is a sealed class - it could be defined in other file/subproject).

(I'm still a bit hesitant to use type resolution, because I feel that it's not widely adopted at the moment, but perhaps I'm wrong. And maybe putting some direct advice + instructions on enabling it in the README would help)

**Comment by BraisGabin** [comment]
Yes, to fix this we will need type resolution.

Right now I'm working on this on the detekt side: 

- https://github.com/detekt/detekt/pull/4960
- https://github.com/detekt/detekt/pull/5130
- https://github.com/detekt/detekt/issues/3577 <- This one is the more interesting one. Once we merge https://github.com/detekt/detekt/pull/5130 I would be able to face that issue. This way if a user has a rule that requires type resolution active but s/he runs detekt without it s/he will get a warning to let s/he know about it.

With this type resolution will not be experimental anymore and we will inform users about the rules that they have active but they aren't working.

**Comment by dimsuz** [comment]
Great work! So I'll get started. Both in adopting type resolution in our work projects and learning to write rules for it (only saw examples in the Detekt repo, didn't try myself).

**Comment by BraisGabin** [comment]
To be honest, they are more complex. But the good part is that there are enoght examples in the detekt repo to find what you need.
